tensorflow/contrib/rnn/kernels/lstm_ops.cc

```diff
@@ -0,0 +1,554 @@
+
+    Tensor* f_tensor = nullptr;
+    OP_REQUIRES_OK(ctx, ctx->allocate_output("f",
+          TensorShape({batch_size, cell_size_}), &f_tensor));
+
+    Tensor* o_tensor = nullptr;
+    OP_REQUIRES_OK(ctx, ctx->allocate_output("o",
+          TensorShape({batch_size, cell_size_}), &o_tensor));
+
+    Tensor* ci_tensor = nullptr;
+    OP_REQUIRES_OK(ctx, ctx->allocate_output("ci",
+          TensorShape({batch_size, cell_size_}), &ci_tensor));
+
+    Tensor* co_tensor = nullptr;
+    OP_REQUIRES_OK(ctx, ctx->allocate_output("co",
+          TensorShape({batch_size, cell_size_}), &co_tensor));
+
+    Tensor* states_tensor = nullptr;
+    OP_REQUIRES_OK(ctx, ctx->allocate_output("states",
+        TensorShape({batch_size, states_size}), &states_tensor));
+
+    Tensor* h_tensor = nullptr;
+    OP_REQUIRES_OK(ctx, ctx->allocate_output("h",
+          TensorShape({batch_size, cell_size_}), &h_tensor));
+
+    // Allocate our temp tensors.
+    Tensor xh_tensor;
+    OP_REQUIRES_OK(ctx, ctx->allocate_temp(DataTypeToEnum<T>::v(),
+        TensorShape({batch_size, input_size + cell_size_}), &xh_tensor));
+
+    Tensor cs_prev_tensor;
+    OP_REQUIRES_OK(ctx, ctx->allocate_temp(DataTypeToEnum<T>::v(),
+        TensorShape({batch_size, cell_size_}), &cs_prev_tensor));
+
+    Tensor h_prev_tensor;
+    OP_REQUIRES_OK(ctx, ctx->allocate_temp(DataTypeToEnum<T>::v(),
+        TensorShape({batch_size, cell_size_}), &h_prev_tensor));
+
+    Tensor icfo_tensor;
+    OP_REQUIRES_OK(ctx, ctx->allocate_temp(DataTypeToEnum<T>::v(),
+        TensorShape({batch_size, cell_size_ * 4}), &icfo_tensor));
+
+    functor::LSTMCellBlockFprop<Device, T, USE_CUBLAS>(
+        batch_size, input_size, cell_size_)(
+        ctx, stream, ctx->eigen_device<Device>(), forget_bias_,
+        x_tensor->matrix<T>(), states_prev_tensor->matrix<T>(),
+        w_tensor->matrix<T>(), b_tensor->vec<T>(), cs_prev_tensor.matrix<T>(),
+        h_prev_tensor.matrix<T>(), xh_tensor.matrix<T>(),
+        i_tensor->matrix<T>(), cs_tensor->matrix<T>(), f_tensor->matrix<T>(),
+        o_tensor->matrix<T>(), ci_tensor->matrix<T>(), co_tensor->matrix<T>(),
+        icfo_tensor.matrix<T>(), states_tensor->matrix<T>(),
+        h_tensor->matrix<T>());
+  }
+
+ private:
+  int64 cell_size_;
+  float forget_bias_;
+};
+
+#define REGISTER_KERNEL(T)                              \
+  REGISTER_KERNEL_BUILDER(Name("LSTMCellBlock")         \
+                              .Device(DEVICE_CPU)       \
+                              .TypeConstraint<T>("T"),  \
+                          LSTMCellBlockOp<CPUDevice, T, false>);
+REGISTER_KERNEL(float);
+REGISTER_KERNEL(double);
+#undef REGISTER_KERNEL
+
+#if GOOGLE_CUDA
+namespace functor {
+#define DECLARE_GPU_SPEC(T)                                       \
+  template <>                                                     \
+  void TensorMemZero<GPUDevice, T>::operator()(                   \
+      const GPUDevice& d, typename TTypes<T>::Vec x);             \
```

Comment:
I removed the memcpy / memzero cause we actually don't use them anymore.
However, I don't think we can use references and pointers in those
functions? (Correct me if I am wrong?). because when we call the functions
we do something like:  func(..., w_tensor->matrix<T>(), ...) unless we do
something like:

auto w_matrix = w_tensor->matrix<T>();

func(.., w_matrix, ...)

do we want to do that instead w/ the extra aliasing? tradeoff of a few more
lines of code for a bit of CPU cycles saved?

--
William Chan
Carnegie Mellon University
(650) 450-9455
williamchan.ca

On Wed, May 11, 2016 at 2:19 PM, ebrevdo <notifications@github.com> wrote:

> In tensorflow/contrib/rnn/kernels/lstm_ops.cc
> <https://github.com/tensorflow/tensorflow/pull/2002#discussion_r62928537>:
>
> > +
> > +#define REGISTER_KERNEL(T)                              \
> > +  REGISTER_KERNEL_BUILDER(Name("LSTMCellBlock")         \
> > +                              .Device(DEVICE_CPU)       \
> > +                              .TypeConstraint<T>("T"),  \
> > +                          LSTMCellBlockOp<CPUDevice, T, false>);
> > +REGISTER_KERNEL(float);
> > +REGISTER_KERNEL(double);
> > +#undef REGISTER_KERNEL
> > +
> > +#if GOOGLE_CUDA
> > +namespace functor {
> > +#define DECLARE_GPU_SPEC(T)                                       \
> > +  template <>                                                     \
> > +  void TensorMemZero<GPUDevice, T>::operator()(                   \
> > +      const GPUDevice& d, typename TTypes<T>::Vec x);             \
>
> you're going to be modifying x, so you should be passing in a pointer.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/2002/files/fc40971bba82762fe413b5e7d6fe12a09722876b#r62928537>
>